man/egpg.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "EGPG" "1" "April 2018" "dashohoxha" "Easy GnuPG"
.
.SH "NAME"
\fBegpg\fR \- Easy GnuPG
.
.SH "SYNOPSIS"
\fBegpg\fR \fICOMMAND\fR [\fIARGS\fR\.\.\.] [\fIOPTIONS\fR\.\.\.]
.
.SH "DESCRIPTION"
There are scads of options presented by GnuPG, which are all part of making it the flexible and powerful an encryption framework that it is\. But it\'s extremely complicated to get started with, and that quite reasonably puts people off\.
.
.P
\fBegpg\fR is a wrapper script that tries to simplify the process of using GnuPG\. In order to simplify things, it is opinionated about the "right" way to use GnuPG\.
.
.SH "COMMANDS"
.
.IP "\(bu" 4
\fBinit\fR [\fIdir\fR]
.
.IP
Initialize egpg\. Optionally give the directory to be used\. If not given, the default directory will be \fI$HOME/\.egpg/\fR
.
.IP "\(bu" 4
\fBmigrate\fR [\-d,\-\-homedir \fIgnupghome\fR]
.
.IP
Get keys and contacts from another gpg directory (by default from \fI$GNUPGHOME\fR)\.
.
.IP "\(bu" 4
[\fBinfo\fR]
.
.IP
Display info about the current configuration and settings\.
.
.IP "\(bu" 4
\fBseal\fR \fIfile\fR [\fIrecipient\fR\.\.\.]
.
.IP
Sign and encrypt a file\. The resulting file will have the extension \fB\.sealed\fR The original file will be erased\.
.
.IP "\(bu" 4
\fBopen\fR \fIfile\.sealed\fR
.
.IP
Decrypt and verify the signature of the given file\. The file has to end with \fB\.sealed\fR and the output will have that extension stripped\.
.
.IP "\(bu" 4
\fBsign\fR \fIfile\fR
.
.IP
Sign a file\. The signature will be saved to <file\fB\.signature\fR>\.
.
.IP "\(bu" 4
\fBverify\fR \fIfile\.signature\fR
.
.IP
Verify the signature\. The signed file must be present as well\.
.
.IP "\(bu" 4
\fBdefault\fR
.
.IP
Use the default GNUPGHOME\.
.
.IP "\(bu" 4
\fBset\fR \fIoption\fR \fIvalue\fR
.
.IP
Change the settings\.
.
.IP "\(bu" 4
\fBkey\fR \fIcommand\fR [\fIoptions\fR\.\.\.]
.
.IP
Commands for handling the key\. For more details see \fBkey help\fR\.
.
.IP "\(bu" 4
\fBcontact\fR \fIcommand\fR [\fIoptions\fR\.\.\.]
.
.IP
Commands for handling the contacts\. For more details see \fBcontact help\fR\.
.
.IP "\(bu" 4
\fBgpg\fR, \fB\-\-\fR \.\.\.
.
.IP
Run any gpg command (but using the configuration settings of egpg)\.
.
.IP "\(bu" 4
\fBhelp\fR
.
.IP
Show this help text\.
.
.IP "\(bu" 4
\fBversion\fR
.
.IP
Show version information\.
.
.IP "" 0
.
.SH "KEY COMMANDS"
.
.IP "\(bu" 4
[\fBls\fR,\fBlist\fR,\fBshow\fR] [\-r,\-\-raw | \-c,\-\-colons] [\-a,\-\-all]
.
.IP
Show the details of the key (optionally in raw format or with colons)\. A list of all the keys can be displayed as well (including the revoked and expired ones)\.
.
.IP "\(bu" 4
\fBgen\fR,\fBgenerate\fR [\fIemail\fR \fIname\fR] [\-n,\-\-no\-passphrase]
.
.IP
Create a new GPG key\. If \fIemail\fR and \fIname\fR are not given as arguments, they will be asked interactively\.
.
.IP "\(bu" 4
\fBfetch\fR [\-d,\-\-homedir \fIgnupghome\fR] [\-k,\-\-key\-id \fIkey\-id\fR]
.
.IP
Get a key from another gpg directory (by default from $GNUPGHOME)\.
.
.IP "\(bu" 4
\fBbackup\fR [\fIkey\-id\fR] [\-q,\-\-qrencode]
.
.IP
Backup key to text file\. If the option \-\-qrencode is given, then a PDF file with 3D barcode will be generated as well\.
.
.IP "\(bu" 4
\fBrestore\fR \fIfile\.tgz\fR
.
.IP
Restore key from file\.
.
.IP "\(bu" 4
\fBsplit\fR [\-d,\-\-dongle \fIdir\fR] [\-b,\-\-backup \fIdir\fR]
.
.IP
Split the key into 3 partial keys and store one of them on the dongle (removable device, usb), keep the other one locally, and use the third one as a backup\. Afterwards, whenever the key needs to be used, the dongle has to be present\.
.
.IP "\(bu" 4
\fBjoin\fR
.
.IP
Join two partial keys into a full key and delete the partials\.
.
.IP "\(bu" 4
\fBrecover\fR \fIbackup\-partial\.key\.xyz\fR
.
.IP
Recover the key from the backup partial key and from the partial key of the home or the dongle\. This is useful when either the dongle or the home partial key is lost\.
.
.IP "\(bu" 4
\fBpass\fR
.
.IP
Change the passphrase\.
.
.IP "\(bu" 4
\fBshare\fR
.
.IP
Publish the key to the keyserver network\.
.
.IP "\(bu" 4
\fBrenew\fR,\fBexpiration\fR [\fIdate\fR]
.
.IP
Renew the key until the given date (by default 1 month from now)\. The \fIdate\fR is in free time format, like "2 months", 2020\-11\-15, "March 7", "5 years" etc\. The date formats are those that are accepted by the command \fBdate \-d\fR (see \fBinfo date\fR)\.
.
.IP "\(bu" 4
\fBrevcert\fR ["description"]
.
.IP
Generate a revocation certificate for the key\.
.
.IP "\(bu" 4
\fBrev\fR,\fBrevoke\fR [\fIrevocation\-certificate\fR]
.
.IP
Cancel the key by publishing the given revocation certificate\.
.
.IP "\(bu" 4
\fBrm\fR,\fBdel\fR,\fBdelete\fR [\fIkey\-id\fR]
.
.IP
Delete the key\.
.
.IP "" 0
.
.SH "CONTACT COMMANDS"
.
.IP "\(bu" 4
\fBls\fR,\fBlist\fR,\fBshow\fR,\fBfind\fR [\fIcontact\fR\.\.\.] [\-r,\-\-raw | \-c,\-\-colons]
.
.IP
Show the details of the contacts (optionally in raw format or with colons)\. A list of all the contacts will be displayed if no one is selected\. A contact can be selected by name, email, id, etc\.
.
.IP "\(bu" 4
\fBrm\fR,\fBdel\fR,\fBdelete\fR \fIcontact\fR\.\.\. [\-f,\-\-force]
.
.IP
Delete the given contact(s)\.
.
.IP "\(bu" 4
\fBexp\fR,\fBexport\fR [\fIcontact\fR\.\.\.] [\-o,\-\-output \fIfile\fR]
.
.IP
Export contact(s) to file\.
.
.IP "\(bu" 4
\fBimp\fR,\fBimport\fR,\fBadd\fR \fIfile\fR
.
.IP
Import (add) contact(s) from file\.
.
.IP "\(bu" 4
\fBfetch\fR [\fIcontact\fR\.\.\.] [\-d,\-\-homedir \fIgnupghome\fR]
.
.IP
Get contacts from another gpg directory (by default from $GNUPGHOME)\.
.
.IP "\(bu" 4
\fBfetch\-uri\fR \fIuri\fR\.\.\.
.
.IP
Retrieve contacts located at the specified URIs\.
.
.IP "\(bu" 4
\fBsearch\fR \fIname\fR [\-s,\-\-keyserver \fIserver\fR]
.
.IP
Search the keyserver network for a person\.
.
.IP "\(bu" 4
\fBreceive\fR,\fBpull\fR \fIcontact\-id\fR [\-s,\-\-keyserver \fIserver\fR]
.
.IP
Download contact from the keyserver network\.
.
.IP "\(bu" 4
\fBcertify\fR \fIcontact\fR [\-p,\-\-publish] [\-l,\-\-level \fIlevel\fR] [\-t,\-\-time \fItime\fR]
.
.IP
You have verified the identity of the contact (the details of the contact, name, email, etc\. are correct and belong to a real person)\. With the \-\-publish option you also share your certification with the world, so that your friends may rely on it if they wish\. The levels of certification are: \fB0\fR (\fBunknown\fR), \fB1\fR (\fBonfaith\fR), \fB2\fR (\fBcasual\fR), \fB3\fR (\fBextensive\fR)\. The time of certification can be: 0 (unlimited), \fIn\fRd (\fIn\fR days), \fIn\fRw (\fIn\fR weeks), \fIn\fRm (\fIn\fR months), \fIn\fRy (\fIn\fR years)\.
.
.IP "\(bu" 4
\fBuncertify\fR \fIcontact\fR
.
.IP
Revoke the certification of a contact\.
.
.IP "\(bu" 4
\fBtrust\fR \fIcontact\fR [\-l,\-\-level \fItrust\-level\fR]
.
.IP
You have verified the identity of the contact and you also trust him to be able to verify correctly and honestly the identities of other people\. The trust levels are: \fB4\fR (\fBfull\fR), \fB3\fR (\fBmarginal\fR), \fB2\fR (\fBnone\fR), \fB1\fR (\fBunknown\fR)
.
.IP "" 0
.
.SH "EXTERNAL COMMANDS"
.
.IP "\(bu" 4
\fBkey fpr\fR
.
.IP
Show the fingerprint of the key\.
.
.IP "\(bu" 4
\fBcontact pick\fR [\-i,\-\-image \fIimagefile\fR]
.
.IP
Read fingerprint as a 2D barcode from camera or from image and receive the given contact from the keyserver network\.
.
.IP "\(bu" 4
\fBkey2dongle\fR [\fIdongle\-dir\fR] [\-r,\-\-reverse]
.
.IP
Move the sign/cert/main key to dongle\. With option \-\-reverse move it back to GNUPGHOME\.
.
.IP "" 0
.
.SH "FILES"
\fB~/\.egpg/\fR The default egpg directory\.
.
.P
\fB~/\.egpg/\.gnupg\fR The GnuPG directory\.
.
.P
\fB~/\.egpg/config\.sh\fR The configuration file\.
.
.P
\fB~/\.egpg/customize\.sh\fR Optional customization file\.
.
.SH "ENVIRONMENT VARIABLES"
\fBEGPG_DIR\fR Overrides the default egpg directory\.
.
.P
\fBGNUPGHOME\fR The directory used by GnuPG\.
.
.SH "CUSTOMIZATION"
The file \fB$EGPG_DIR/customize\.sh\fR can be used to redefine and customize some functions, without having to touch the code of the main script\. Also, external commands can be customized (or new commands can be defined) by adding the file \fB$EGPG_DIR/cmd_command\.sh\fR, which contains the function \fBcmd_command() { \. \. \. }\fR\.
.
.P
In general, for an external command the script will first look for \fB$EGPG_DIR/cmd_command\.sh\fR, then for \fB$LIB/ext/$PLATFORM/cmd_command\.sh\fR, and finally for \fB$LIB/ext/cmd_command\.sh\fR\. The first that is found is loaded and used\. For key commands the name of the file must be \fBcmd_key_command\.sh\fR and for contact commands it must be \fBcmd_contact_command\.sh\fR\.
.
.SH "SIMPLE EXAMPLE"
Some basic usage is demonstrated on the examples below\.
.
.P
testuser@laptop:~$ \fBrm \-rf ~/\.egpg/\fR
.
.P
testuser@laptop:~$ \fBkillall \-u $USER gpg\-agent\fR
.
.P
testuser@laptop:~$ \fBegpg\fR
.
.IP "" 4
.
.nf

No directory \'/home/testuser/\.egpg\'
Try first: egpg init
.
.fi
.
.IP "" 0
.
.P
testuser@laptop:~$ \fBegpg init\fR
.
.IP "" 4
.
.nf

mkdir: created directory \'/home/testuser/\.egpg\'

Appended the following lines to \'/home/testuser/\.bashrc\':
\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-8<\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
### start egpg config
export GPG_TTY=$(tty)
export EGPG_DIR="/home/testuser/\.egpg"
#export GNUPGHOME="/home/testuser/\.egpg/\.gnupg"
### end egpg config
\-\-\-\-\-\-\-\-\-\-\-\-\-\-\->8\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
Please reload it to enable the new config:
    source "/home/testuser/\.bashrc"
.
.fi
.
.IP "" 0
.
.P
testuser@laptop:~$ \fBsource "/home/testuser/\.bashrc"\fR
.
.P
testuser@laptop:~$ \fBegpg\fR
.
.IP "" 4
.
.nf

EasyGnuPG 2\.2\-1\.0    ( https://github\.com/easygnupg/egpg )

EGPG_DIR="/home/testuser/\.egpg"
GNUPGHOME="/home/testuser/\.egpg/\.gnupg"
DONGLE=""
KEYSERVER="hkp://keys\.gnupg\.net"
GPG_TTY="/dev/pts/18"
SHARE=no
DEBUG=no

No valid key found\.

Try first:  egpg key gen
       or:  egpg key fetch
       or:  egpg key restore
       or:  egpg key recover
.
.fi
.
.IP "" 0
.
.P
testuser@laptop:~$ \fBegpg key gen test@example\.org "Test User"\fR
.
.IP "" 4
.
.nf

Creating a new key\.

Enter passphrase for the new key:
Retype the passphrase of the key:
Please enter the passphrase to unlock the OpenPGP secret key:
"Test User <test@example\.org>"
4096\-bit RSA key, ID 01D532A283DC1CBF,
created 2016\-06\-06\.

Passphrase:

id: 01D532A283DC1CBF
uid: Test User <test@example\.org>
fpr: 4E91 DD0E EDD7 49F2 B6DE D9CE 01D5 32A2 83DC 1CBF
trust: ultimate
sign: 01D532A283DC1CBF 2016\-06\-06 2016\-07\-06
decr: 7BD4B1BE5D76CC0D 2016\-06\-06 2016\-07\-06

Revocation certificate is at:
    "/home/testuser/\.egpg/\.gnupg/openpgp\-revocs\.d/4E91DD0EEDD749F2B6DED9CE01D532A283DC1CBF\.rev"
    "/home/testuser/\.egpg/\.gnupg/openpgp\-revocs\.d/4E91DD0EEDD749F2B6DED9CE01D532A283DC1CBF\.rev\.pdf"
.
.fi
.
.IP "" 0
.
.P
testuser@laptop:~$ \fBegpg info\fR
.
.IP "" 4
.
.nf

EasyGnuPG 2\.2\-1\.0    ( https://github\.com/easygnupg/egpg )

EGPG_DIR="/home/testuser/\.egpg"
GNUPGHOME="/home/testuser/\.egpg/\.gnupg"
DONGLE=""
KEYSERVER="hkp://keys\.gnupg\.net"
GPG_TTY="/dev/pts/18"
SHARE=no
DEBUG=no

id: 01D532A283DC1CBF
uid: Test User <test@example\.org>
fpr: 4E91 DD0E EDD7 49F2 B6DE D9CE 01D5 32A2 83DC 1CBF
trust: ultimate
sign: 01D532A283DC1CBF 2016\-06\-06 2016\-07\-06
decr: 7BD4B1BE5D76CC0D 2016\-06\-06 2016\-07\-06
.
.fi
.
.IP "" 0
.
.P
testuser@laptop:~$ \fBegpg key\fR
.
.IP "" 4
.
.nf

id: 01D532A283DC1CBF
uid: Test User <test@example\.org>
fpr: 4E91 DD0E EDD7 49F2 B6DE D9CE 01D5 32A2 83DC 1CBF
trust: ultimate
sign: 01D532A283DC1CBF 2016\-06\-06 2016\-07\-06
decr: 7BD4B1BE5D76CC0D 2016\-06\-06 2016\-07\-06
.
.fi
.
.IP "" 0
.
.P
testuser@laptop:~$ \fBegpg key fpr\fR
.
.IP "" 4
.
.nf

4E91 DD0E EDD7 49F2 B6DE D9CE 01D5 32A2 83DC 1CBF
.
.fi
.
.IP "" 0
.
.P
testuser@laptop:~$ \fBegpg key renew 1 year\fR
.
.IP "" 4
.
.nf

Please enter the passphrase to unlock the OpenPGP secret key:
"Test User <test@example\.org>"
4096\-bit RSA key, ID 01D532A283DC1CBF,
created 2016\-06\-06\.

Passphrase:

id: 01D532A283DC1CBF
uid: Test User <test@example\.org>
fpr: 4E91 DD0E EDD7 49F2 B6DE D9CE 01D5 32A2 83DC 1CBF
trust: ultimate
sign: 01D532A283DC1CBF 2016\-06\-06 2017\-06\-06
decr: 7BD4B1BE5D76CC0D 2016\-06\-06 2017\-06\-06
.
.fi
.
.IP "" 0
.
.P
testuser@laptop:~$ \fBecho "This is a test message\." > test\.txt\fR
.
.P
testuser@laptop:~$ \fBegpg seal test\.txt\fR
.
.IP "" 4
.
.nf

Please enter the passphrase to unlock the OpenPGP secret key:
"Test User <test@example\.org>"
4096\-bit RSA key, ID 01D532A283DC1CBF,
created 2016\-06\-06\.

Passphrase:
.
.fi
.
.IP "" 0
.
.P
testuser@laptop:~$ \fBegpg open test\.txt\.sealed\fR
.
.IP "" 4
.
.nf

Please enter the passphrase to unlock the OpenPGP secret key:
"Test User <test@example\.org>"
4096\-bit RSA key, ID 7BD4B1BE5D76CC0D,
created 2016\-06\-06 (main key ID 01D532A283DC1CBF)\.

Passphrase:
gpg: Signature made Mon 06 Jun 2016 06:28:52 PM CEST
gpg:                using RSA key 01D532A283DC1CBF
gpg: Good signature from "Test User <test@example\.org>" [ultimate]
.
.fi
.
.IP "" 0
.
.P
testuser@laptop:~$ \fBegpg sign test\.txt\fR
.
.IP "" 4
.
.nf

Please enter the passphrase to unlock the OpenPGP secret key:
"Test User <test@example\.org>"
4096\-bit RSA key, ID 01D532A283DC1CBF,
created 2016\-06\-06\.

Passphrase:
.
.fi
.
.IP "" 0
.
.P
testuser@laptop:~$ \fBegpg verify test\.txt\.signature\fR
.
.IP "" 4
.
.nf

gpg: Signature made Mon 06 Jun 2016 06:32:23 PM CEST
gpg:                using RSA key 01D532A283DC1CBF
gpg: Good signature from "Test User <test@example\.org>" [ultimate]
.
.fi
.
.IP "" 0
.
.P
testuser@laptop:~$ \fBegpg contact search Dashamir\fR
.
.IP "" 4
.
.nf

gpg: data source: http://svcs4\.riverwillow\.net\.au:11371
(1) Dashamir Hoxha <dashohoxha@gmail\.com>
      2048 bit RSA key 562AC309C01D2DBD, created: 2015\-12\-27, expires: 2020\-12\-31
(2) Dashamir Hoxha <dashohoxha@gmail\.com>
    keybase\.io/dashohoxha <dashohoxha@keybase\.io>
      4096 bit RSA key 0967FD258D6414F9, created: 2015\-05\-27, expires: 2017\-01\-05
(3) Dashamir Hoxha <dashohoxha@gmail\.com>
    Dashamir Hoxha <d_hoxha@dogana\.gov\.al>
      2048 bit RSA key FD06AA8E55D59B28, created: 2010\-12\-12, expires: 2015\-12\-13 (expired)
Keys 1\-3 of 3 for "Dashamir"\.  Enter number(s), N)ext, or Q)uit > 1
.
.fi
.
.IP "" 0
.
.P
testuser@laptop:~$ \fBegpg contact ls\fR
.
.IP "" 4
.
.nf

id: 01D532A283DC1CBF
uid: Test User <test@example\.org>
fpr: 4E91 DD0E EDD7 49F2 B6DE D9CE 01D5 32A2 83DC 1CBF
trust: ultimate
sign: 01D532A283DC1CBF 2016\-06\-06 2017\-06\-06
decr: 7BD4B1BE5D76CC0D 2016\-06\-06 2017\-06\-06


id: 562AC309C01D2DBD
uid: Dashamir Hoxha <dashohoxha@gmail\.com>
fpr: 1EC0 8B86 1350 EE19 8053 7941 562A C309 C01D 2DBD
sign: 562AC309C01D2DBD 2015\-12\-27 2020\-12\-31
decr: 95B9FAD9DEF9A02A 2015\-12\-27 2020\-12\-31
.
.fi
.
.IP "" 0
.
.P
testuser@laptop:~$ \fBegpg contact certify Dashamir\fR
.
.IP "" 4
.
.nf

Please enter the passphrase to unlock the OpenPGP secret key:
"Test User <test@example\.org>"
4096\-bit RSA key, ID 01D532A283DC1CBF,
created 2016\-06\-06\.

Passphrase:

pub  rsa2048/562AC309C01D2DBD
     created: 2015\-12\-27  expires: 2020\-12\-31  usage: SC
     trust: unknown       validity: unknown
 Primary key fingerprint: 1EC0 8B86 1350 EE19 8053  7941 562A C309 C01D 2DBD

     Dashamir Hoxha <dashohoxha@gmail\.com>

This key is due to expire on 2020\-12\-31\.

I have checked this key casually\.
.
.fi
.
.IP "" 0
.
.P
testuser@laptop:~$ \fBegpg contact trust Dashamir\fR
.
.IP "" 4
.
.nf

uid: Dashamir Hoxha <dashohoxha@gmail\.com>
trust: marginal
.
.fi
.
.IP "" 0
.
.P
testuser@laptop:~$ \fBegpg contact ls Dashamir\fR
.
.IP "" 4
.
.nf

id: 562AC309C01D2DBD
uid: Dashamir Hoxha <dashohoxha@gmail\.com>
fpr: 1EC0 8B86 1350 EE19 8053 7941 562A C309 C01D 2DBD
trust: marginal
sign: 562AC309C01D2DBD 2015\-12\-27 2020\-12\-31
decr: 95B9FAD9DEF9A02A 2015\-12\-27 2020\-12\-31
certified by: Test User <test@example\.org> (01D532A283DC1CBF)
.
.fi
.
.IP "" 0
.
.P
testuser@laptop:~$ \fBegpg contact rm Dashamir\fR
.
.IP "" 4
.
.nf

gpg (GnuPG) 2\.1\.11; Copyright (C) 2016 Free Software Foundation, Inc\.
This is free software: you are free to change and redistribute it\.
There is NO WARRANTY, to the extent permitted by law\.

pub  rsa2048/562AC309C01D2DBD 2015\-12\-27 Dashamir Hoxha <dashohoxha@gmail\.com>

Delete this key from the keyring? (y/N) y
.
.fi
.
.IP "" 0
.
.P
testuser@laptop:~$ \fBegpg key revoke\fR
.
.IP "" 4
.
.nf

Revocation will make your current key useless\.
You\'ll need to generate a new one\.
Are you sure about this? [y/N] y
.
.fi
.
.IP "" 0
.
.P
testuser@laptop:~$ \fBegpg key ls\fR
.
.IP "" 4
.
.nf

No valid key found\.

Try first:  egpg key gen
       or:  egpg key fetch
       or:  egpg key restore
       or:  egpg key recover
.
.fi
.
.IP "" 0
.
.P
testuser@laptop:~$ \fBrm \-rf ~/\.egpg/\fR
.
.SH "ADVANCED EXAMPLE"
.
.SS "Get the key and contacts from an existing $GNUPGHOME:"
testuser@laptop:~$ \fBrm \-rf ~/\.egpg/\fR
.
.P
testuser@laptop:~$ \fBkillall \-u $USER gpg\-agent\fR
.
.P
testuser@laptop:~$ \fBegpg init\fR
.
.IP "" 4
.
.nf

mkdir: created directory \'/home/testuser/\.egpg\'

Appended the following lines to \'/home/testuser/\.bashrc\':
\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-8<\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
### start egpg config
export GPG_TTY=$(tty)
export EGPG_DIR="/home/testuser/\.egpg"
#export GNUPGHOME="/home/testuser/\.egpg/\.gnupg"
### end egpg config
\-\-\-\-\-\-\-\-\-\-\-\-\-\-\->8\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
Please reload it to enable the new config:
    source "/home/testuser/\.bashrc"
.
.fi
.
.IP "" 0
.
.P
testuser@laptop:~$ \fBsource "/home/testuser/\.bashrc"\fR
.
.P
testuser@laptop:~$ \fBegpg migrate\fR
.
.IP "" 4
.
.nf

Importing key from: /home/testuser/\.gnupg

Importing contacts from: /home/testuser/\.gnupg
.
.fi
.
.IP "" 0
.
.SS "Change the passphrase of the key:"
testuser@laptop:~$ \fBegpg key pass\fR
.
.IP "" 4
.
.nf

Please enter the passphrase to unlock the OpenPGP secret key:
"Test 1 <test1@exampel\.org>"
4096\-bit RSA key, ID DB7065FD432FF513,
created 2016\-06\-06\.

Passphrase:
Please enter the new passphrase
Passphrase:
Repeat:
.
.fi
.
.IP "" 0
.
.SS "Use the default GNUPGHOME:"
testuser@laptop:~$ \fBegpg default\fR
.
.IP "" 4
.
.nf

\'/home/testuser/\.gnupg\' \-> \'/home/testuser/\.gnupg\-old\'
\'/home/testuser/\.egpg/\.gnupg\' \-> \'/home/testuser/\.gnupg\'
OK
.
.fi
.
.IP "" 0
.
.P
Afterwards \fBegpg\fR will use the directory \fB~/\.gnupg\fR (or whatever is in the environment variable $GNUPGHOME) for the key and contacts\.
.

src/egpg.sh
#!/usr/bin/env bash
#
# EasyGnuPG is a wrapper around GnuPG to simplify its operations.
# Copyright (C) 2016  Dashamir Hoxha
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see http://www.gnu.org/licenses/

umask 077
set -o pipefail

VERSION="2.2-1.0"

LIBDIR="$(dirname "$0")"

# make sure that these global variables
# do not inherit values from the environment
unset WORKDIR GPG_KEY FPR

source "$LIBDIR/auxiliary.sh"
source "$LIBDIR/platform.sh"

cmd_version() {
    echo
    echo "EasyGnuPG $VERSION    ( https://github.com/easygnupg/egpg )"
    echo
}

cmd() {
    PROGRAM="${0##*/}"
    local cmd="$1" ; shift
    case "$cmd" in
        migrate)          call cmd_migrate "$@" ;;
        ''|info)          call cmd_info "$@" ;;
        seal)             call cmd_seal "$@" ;;
        open)             call cmd_open "$@" ;;
        sign)             call cmd_sign "$@" ;;
        verify)           call cmd_verify "$@" ;;
        set)              call cmd_set "$@" ;;
        default)          call cmd_default "$@" ;;

        --|gpg)           cmd_gpg "$@" ;;
        key)              cmd_key "$@" ;;
        c|contact)        cmd_contact "$@" ;;

        ext)              cmd_ext_help "$@" ;;
        *)                call_ext cmd_$cmd "$@" ;;
    esac
}

cmd_gpg() { gpg "$@"; }

cmd_key() {
    local cmd="$1" ; shift
    case "$cmd" in
        help)             call cmd_key_help "$@" ;;
        gen|generate)     call cmd_key_gen "$@" ;;
        ''|ls|list|show)  call cmd_key_list "$@" ;;
        rm|del|delete)    call cmd_key_delete "$@" ;;
        backup)           call cmd_key_backup "$@" ;;
        restore)          call cmd_key_restore "$@" ;;
        fetch)            call cmd_key_fetch "$@" ;;
        renew|expiration) call cmd_key_renew "$@" ;;
        share)            call cmd_key_share "$@" ;;
        revcert)          call cmd_key_revcert "$@" ;;
        rev|revoke)       call cmd_key_rev "$@" ;;
        pass)             call cmd_key_pass "$@" ;;
        split)            call cmd_key_split "$@" ;;
        join)             call cmd_key_join "$@" ;;
        recover)          call cmd_key_recover "$@" ;;
        help)             call cmd_key_help "$@" ;;
        *)                call_ext cmd_key_$cmd "$@" ;;
    esac
}

cmd_contact() {
    local cmd="$1" ; shift
    case "$cmd" in
        ''|help)          call cmd_contact_help "$@" ;;
        exp|export)       call cmd_contact_export "$@" ;;
        imp|import|add)   call cmd_contact_import "$@" ;;
        fetch)            call cmd_contact_fetch "$@" ;;
        fetch-uri)        call cmd_contact_fetchuri "$@" ;;
        search)           call cmd_contact_search "$@" ;;
        receive|pull)     call cmd_contact_receive "$@" ;;
        ls|list|show|find)call cmd_contact_list "$@" ;;
        rm|del|delete)    call cmd_contact_delete "$@" ;;
        certify)          call cmd_contact_certify "$@" ;;
        uncertify)        call cmd_contact_uncertify "$@" ;;
        trust)            call cmd_contact_trust "$@" ;;
        *)                call_ext cmd_contact_$cmd "$@" ;;
    esac
}

call() {
    local cmd=$1; shift
    local file="$LIBDIR/$(echo $cmd | tr _ /).sh"
    [[ -f "$file" ]] || fail "Cannot find command file: $file"
    source "$file"
    if [[ $1 == "help" ]]
    then ${cmd}_help
    else $cmd "$@"
    fi
}

call_fn() {
    local fn=$1; shift
    local file="$LIBDIR/fn/$fn.sh"
    [[ -f "$file" ]] || fail "Cannot find function file: $file"
    source "$file"
    $fn "$@"
}

call_ext() {
    local cmd=$1; shift

    load() {
        debug loading: "$1"
        source "$1"
    }

    if   [[ -f "$EGPG_DIR/$cmd.sh" ]];             then load "$EGPG_DIR/$cmd.sh"
    elif [[ -f "$LIBDIR/ext/$PLATFORM/$cmd.sh" ]]; then load "$LIBDIR/ext/$PLATFORM/$cmd.sh"
    elif [[ -f "$LIBDIR/ext/$cmd.sh" ]];           then load "$LIBDIR/ext/$cmd.sh"
    else
        echo -e "Unknown command '$cmd'.\nTry:  $(basename $0) help"
        return 1
    fi

    debug running: $cmd "$@"
    if [[ $1 == 'help' ]]
    then ${cmd}_help
    else $cmd "$@"
    fi
}

cmd_ext_help() {
    cat <<-_EOF
An external 'command' can be loaded from:
    1. $EGPG_DIR/cmd_command.sh
    2. $LIBDIR/ext/$PLATFORM/cmd_command.sh
    3. $LIBDIR/ext/cmd_command.sh
The first file that is found is loaded and used.
For key commands the name of the file must be: cmd_key_command.sh
and for contact commands it must be: cmd_contact_command.sh

_EOF

    for cmd_file in \
        "$LIBDIR/ext"/cmd_*.sh \
        "$LIBDIR/ext/$PLATFORM"/cmd_*.sh \
        "$EGPG_DIR"/cmd_*.sh
    do
        [[ -f "$cmd_file" ]] || continue
        source "$cmd_file"
        cmd=$(basename "${cmd_file%%.sh}")
        [[ $cmd == *"test" ]] && continue
        ${cmd}_help
    done
}

config() {
    # read the config file
    local config_file="$EGPG_DIR/config.sh"
    ENV_GNUPGHOME=${GNUPGHOME:-~/.gnupg}
    unset SHARE KEYSERVER GNUPGHOME DONGLE DEBUG
    [[ -f "$config_file" ]] && source "$config_file"

    # set defaults, if some configurations are missing
    SHARE=${SHARE:-no}
    KEYSERVER=${KEYSERVER:-hkp://keys.gnupg.net}
    GNUPGHOME=${GNUPGHOME:-$EGPG_DIR/.gnupg}
    [[ "$GNUPGHOME" == "default" ]] && GNUPGHOME="$ENV_GNUPGHOME"
    DEBUG=${DEBUG:-no}

    export GNUPGHOME
    export GPG_TTY=$(tty)

    # create the config file, if it does not exist
    local gpghome="$GNUPGHOME"
    [[ "GNUPGHOME" == "ENV_GNUPGHOME" ]] && gpghome='default'
    [[ -f "$config_file" ]] || cat <<-_EOF > "$config_file"
# If yes, push local changes to the keyserver network.
SHARE=$SHARE
KEYSERVER="$KEYSERVER"

# GPG homedir to be used. If "default", then use the default one,
# (whatever is in the environment $GNUPGHOME, usually ~/.gnupg).
GNUPGHOME="$gpghome"

# Path of the dongle.
DONGLE="$DONGLE"

# If true, print debug output.
DEBUG=$DEBUG
_EOF
}

main() {
    local gnupg_version=$(gpg_version)
    [[ ${gnupg_version%.*} == "2.1" ]] || fail "These scripts are supposed to work with GnuPG 2.1"

    # handle some basic commands
    case "$1" in
        v|-v|version|--version)  shift; cmd_version "$@" ; exit 0 ;;
        help|-h|--help)          shift; call cmd_help "$@" ; exit 0 ;;
        init)                    shift; call cmd_init "$@" ; exit 0 ;;
    esac

    # set config variables
    export EGPG_DIR="${EGPG_DIR:-$HOME/.egpg}"
    [[ -d "$EGPG_DIR" ]] || fail "No directory '$EGPG_DIR'\nTry first: $(basename $0) init"
    config

    # customize platform dependent functions
    PLATFORM="$(uname | cut -d _ -f 1 | tr '[:upper:]' '[:lower:]')"
    local platform_file="$LIBDIR/platform/$PLATFORM.sh"
    [[ -f "$platform_file" ]] && source "$platform_file"

    # The file 'customize.sh' can be used to redefine
    # and customize some functions, without having to
    # touch the code of the main script.
    local customize_file="$EGPG_DIR/customize.sh"
    [[ -f "$customize_file" ]] && source "$customize_file"

    # run the command
    cmd "$@"
}

main "$@"
